Graph function tweaks

eperim opened the pull request:
This should improve performance of most functions in misc.jl. More notably, for computing the mean shortest path we replace Djikstra's algorithm with Floyd-Warshall's.

eperim committed:
Tweak misc.jl functions for improved performance

eperim committed:
Replace Djikstra's algorithm with Floyd-Marshall's

eperim committed:
Change CI scripts

eperim committed:
Pin numpy version in build.jl

eperim committed:
Update documentation

rofinn commented on .travis.yml:
```diff
@@ -34,7 +34,7 @@ before_install:
   - export PYTHON=""  # Configure PyCall to use the Conda.jl package's Python
   - OPTIONS=$(julia -e 'VERSION == v"0.6.0" && print("--inline=no")')  # https://github.com/JuliaLang/julia/issues/22582
-  - julia $OPTIONS -e 'Pkg.clone(pwd()); Pkg.build("HDF5"); Pkg.build("SyntheticGrids"); Pkg.checkout("PyCall"); Pkg.test("SyntheticGrids"; coverage=true)'
+  - julia $OPTIONS -e 'Pkg.clone(pwd()); Pkg.add("HDF5"); Pkg.build("HDF5"); Pkg.build("SyntheticGrids"); Pkg.checkout("PyCall"); Pkg.test("SyntheticGrids"; coverage=true)'
```
Can't you just put HDF5 in the REQUIRE file?

rofinn commented on deps/build.jl:
```diff
@@ -26,6 +26,12 @@ Conda.add("llvmlite==$LLVMLITE_VERSION")  # https://github.com/numba/llvmlite
+open(Pkg.dir("Conda", "deps", "usr", "conda-meta", "pinned"), "a") do f
+    println(f, "numpy 1.12.1")
+end
+Conda.update()
```
I'd suggest moving the `Conda.update()` above the `Conda.add("pandapower")` and just include `Conda.add("numpy==1.12.1")`.

rofinn commented on src/misc.jl:
```diff
@@ -11,13 +11,12 @@ count_bus_type(grid::Grid) = (nload(buses(grid)), ngen(buses(grid)))
 # effectively giving us two representations of the network (fine and coarse)
 """
-    laplacian(con_mat::AbstractMatrix{<:Real})
+    laplacian(con_mat::AbstractMatrix)
```
Could probably keep the `<:Real`.

rofinn commented on src/misc.jl:
```diff
@@ -30,31 +29,53 @@ function laplacian(con_mat::AbstractMatrix{<:Real})
     return Symmetric(lap, :L)
 end
 
+function laplacian(con_mat::AbstractMatrix{Int})
```
Since you're already special casing the `Bool` case:

```julia
function laplacian(con_mat::AbstractMatrix{T}) where T<:Real
    return laplacian(con_mat .> zero(T)) # treating the substation graph as a simple graph.
end
```

rofinn commented on src/misc.jl:
```diff
@@ -30,31 +29,53 @@ function laplacian(con_mat::AbstractMatrix{<:Real})
+Return the total number of links in a system with connectivity matrix `con_mat`. The graph
+will be treated as a simple graph.
+"""
+function total_links(con_mat::AbstractMatrix{Int})
```
Same issue. You can still use `<:Real`, but just use `zero(T)` in the comparison.

rofinn commented on src/misc.jl:
```diff
@@ -152,40 +175,101 @@ function dijkstra(con_mat::AbstractMatrix{<:Real}, v1)
+Compute minimum distances to each node in the network using Floyd-Warshall's algorithm. Here
+we assume that the weigths matrix is symmetric.
+"""
+function floyd_warshall(weights::AbstractMatrix{<:Real})
```
http://juliagraphs.github.io/LightGraphs.jl/latest/pathing.html#LightGraphs.floyd_warshall_shortest_paths

eperim replied:
I think this goes back to that discussion about wether we should add an extra dependency for a single method (that is not something really important in this package). Also, this implementation has a decent performance.

rofinn replied:
Well `dijkstra` was already used prior to this. I think the number of performant algorithms and [integration](http://juliagraphs.github.io/LightGraphs.jl/latest/integration.html) with other packages is a reasonable argument for adding the extra dependency. I'll note that the dependency stack for pandapower is much more complex (e.g., Conda.jl, PyCall.jl, pandapower w/ specific python dependency versions).

eperim replied:
The pandapower thing is very true. `dijkstra` is not being used anymore, and in this one we actually have comparable performance:
```
julia> using SyntheticGrids
WARNING: `SyntheticGrids.jl may not run properly if inlinning is enabled. Please start Julia with \`--inline=no\``
WARNING: Method definition ==(Base.Nullable{S}, Base.Nullable{T}) in module Base at nullable.jl:238 overwritten in module NullableArrays at /Users/ericperim/.julia/v0.6/NullableArrays/src/operators.jl:99.

julia> grid = Grid(false); nothing

julia> using LightGraphs
WARNING: using LightGraphs.Grid in module Main conflicts with an existing identifier.

julia> graph = LightGraphs.Graph(adjacency(grid))
{137, 167} undirected simple Int64 graph

julia> function mean_short_path_lg(graph::LightGraphs.Graph)
           fws = LightGraphs.floyd_warshall_shortest_paths(graph)
           n = size(fws.dists, 1)
           return sum(fws.dists)/(n * n)
       end
mean_short_path_lg (generic function with 1 method)

julia> for i in 1:1000
           el_time[1] += @elapsed SyntheticGrids.mean_shortest_path(adjacency(grid))
           el_time[2] += @elapsed mean_short_path_lg(graph)
       end

julia> println(el_time ./ 1000)
[0.00379599, 0.00483175]
```
of course LightGraphs offers way more functionality, but we are not using them, and that comes at the cost of dealing more types.

PS: the direct comparison between the times here is also not fair because the SyntheticGrids version makes extra assumptions about the graph that are always true when working with electricity grids.

rofinn replied:
> cost of dealing more types

By this do you mean that you don't think the 1 extra layer of abstraction isn't worth it? Wouldn't the `LightGraphs.Graph` type just replace the adjacency matrix?

eperim replied:
I haven't really looked much into LightGraphs, so maybe my laziness is clouding my judgement here, but I get the impression that several of their methods use custom return types (for instance `floyd_warshall_shortest_paths` returns a `FloydWarshallState` and `dijkstra_shortest_paths` returns a `DijkstraState`), it is not just one extra type (Graph). I am sure that makes sense when you are using further functionalities, but for us here we just really need a simple matrix. On top of that, with our further assumptions we end up with a slightly better performance on this specific task as shown above.
As for the adjacency matrix, it is a necessary piece of the `SyntheticGrids.Grid` object, so it will already be there.

The reason for having pandapower as a part of the stack is that it runs OPF, which is something that we cannot quickly add to the package, while the mean shortest path calculation was simple to add.

Basically, I don't think it is worth integrating with another package just for this one function, but, if you believe we should have all kinds of graph tools at our disposal, then I can go over LightGraphs later and write a proper API to integrate SyntheticGrids with it (for another PR).

rofinn replied:
> for instance floyd_warshall_shortest_paths returns a FloydWarshallState and dijkstra_shortest_paths returns a DijkstraState

FWIW, both those types have a `dists` field that you can extract without really caring about what the type is, but I get your point.

Okay, for now I'd probably suggest opening an issue to integrate with LightGraphs.jl. If we identify more situations that would benefit from using LightGraphs.jl we can keep track of them there.

eperim committed:
Move HDF5 to REQUIRE

eperim committed:
Change Conda commands in build.jl

eperim committed:
Change type annotations in misc.jl

eperim committed:
Change test file deletion order

Appveyor's CI sometimes runs into permission issues when trying to
delete test files. This is an issue with Windows not freeing the
files after they've already been used and closed. Just moving the
delete command around tends to solve this.

eperim committed:
Solve Appveyor testing issues
